File: scripts/handoff.py
"""Build an Orbit handoff pack. Never injects into a live TUI."""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(Path(__file__).resolve().parent))
from snapshot import UUID_RE, build, grok_home, redact  # noqa: E402
from web_adapters import status as web_status  # noqa: E402
from projects import resolve_clone  # noqa: E402


def git_head(cwd: str) -> str | None:
    if not cwd or not Path(cwd).is_dir():
        return None
    try:
        branch = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            cwd=cwd,
            text=True,
            timeout=2,
            stderr=subprocess.DEVNULL,
        ).strip()
        sha = subprocess.check_output(
            ["git", "rev-parse", "--short", "HEAD"],
            cwd=cwd,
            text=True,
            timeout=2,
            stderr=subprocess.DEVNULL,
        ).strip()
    except Exception:
        return None
    if not branch:
        return None
    return f"{branch} @ {sha}"


def acp_cwd_for(session: dict) -> str:
    clone = resolve_clone(session)
    if clone:
        return clone
    sid = str(session.get("id") or "")
    src = str(session.get("source") or "")
    if sid.startswith("web:") or src in {"cursor_web", "grok_web"}:
        return ""
    cwd = session.get("cwd") or ""
    if cwd and Path(cwd).is_dir():
        return cwd
    return os.environ.get("USERPROFILE") or os.environ.get("HOME") or ""


def live_ids() -> set[str]:
    path = grok_home() / "active_sessions.json"
    try:
        rows = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return set()
    from snapshot import pid_alive

    out = set()
    for row in rows if isinstance(rows, list) else []:
        sid = str(row.get("session_id") or "")
        pid = int(row.get("pid") or 0)
        if UUID_RE.match(sid) and pid_alive(pid):
            out.add(sid)
    return out


def find_session(sid: str) -> dict | None:
    snap = build()
    hit = next((s for s in snap.get("sessions") or [] if s.get("id") == sid), None)
    if hit:
        return hit
    if str(sid).startswith("web:"):
        st = web_status()
        hit = next((s for s in st.get("sessions") or [] if s.get("id") == sid), None)
        if hit:
            from projects import assign_slug, build_catalog
            from pr_state import apply_to_sessions

            hit["project_id"] = assign_slug(hit, build_catalog())
            apply_to_sessions([hit])
        return hit
    return None


def tail_events(session: dict) -> list[dict]:
    disk = session.get("disk_path")
    cwd = session.get("cwd") or ""
    sid = session.get("id") or ""
    from snapshot import encode_cwd, tail_events as _tail

    path = None
    if disk:
        path = Path(disk) / "updates.jsonl"
    elif cwd and UUID_RE.match(str(sid)):
        path = grok_home() / "sessions" / encode_cwd(cwd) / sid / "updates.jsonl"
    if path is None:
        return []
    return _tail(path, 24000, 12)


def plan_excerpt(session: dict) -> str:
    disk = session.get("disk_path")
    cwd = session.get("cwd") or ""
    sid = session.get("id") or ""
    from snapshot import encode_cwd

    dirs = []
    if disk:
        dirs.append(Path(disk))
    if cwd and UUID_RE.match(str(sid)):
        dirs.append(grok_home() / "sessions" / encode_cwd(cwd) / sid)
    for d in dirs:
        for name in ("plan.md", "plan.json"):
            p = d / name
            if p.exists():
                try:
                    return redact(p.read_text(encoding="utf-8", errors="replace"))[:1200]
                except OSError:
                    return ""
    return ""


def _append_well(lines: list[str], session: dict, sid: str) -> None:
    from projects import is_named_well, member_kind

    pid = str(session.get("project_id") or "")
    if not is_named_well(pid):
        return
    try:
        snap = build()
    except Exception:
        return
    for row in snap.get("attention") or []:
        title = str(row.get("title") or "")
        if row.get("kind") == "desk_claim" and f"claim {pid}" in title.lower():
            lines.append(f"desk: {redact(title)}")
            break
    members = [
        s
        for s in snap.get("sessions") or []
        if s.get("id") != sid and str(s.get("project_id") or "") == pid
    ][:6]
    if not members:
        return
    lines.extend(["", "Well", "----"])
    for row in members:
        lines.append(f"- {redact(row.get('title') or row.get('id'))} ({member_kind(row)})")


def build_handoff(sid: str) -> dict:
    sid = (sid or "").strip()
    if not (UUID_RE.match(sid) or sid.startswith("web:")):
        return {"error": "invalid session id"}
    session = find_session(sid)
    if not session:
        return {"error": "session not in snapshot"}
    live = bool(session.get("live")) or sid in live_ids()
    cwd = session.get("cwd") or ""
    acp_cwd = acp_cwd_for(session)
    branch = git_head(acp_cwd or cwd)
    events = tail_events(session)
    plan = plan_excerpt(session)
    lines = [
        "ORBIT HANDOFF",
        f"source: {session.get('source') or 'unknown'}",
        f"id: {sid}",
        f"title: {redact(session.get('title') or sid)}",
    ]
    if session.get("project_id"):
        lines.append(f"project: {session['project_id']}")
    if cwd:
        lines.append(f"cwd: {cwd}")
    if acp_cwd and acp_cwd != cwd:
        lines.append(f"clone: {acp_cwd}")
    if session.get("remote"):
        lines.append(f"remote: {session['remote']}")
    if branch:
        lines.append(f"branch: {branch}")
    if session.get("model"):
        lines.append(f"model: {session['model']}")
    if session.get("url"):
        lines.append(f"url: {session['url']}")
    if session.get("pr_url"):
        st = session.get("pr_state") or "unknown"
        lines.append(f"pr: {st} {session['pr_url']}")
        if session.get("pr_file_count") is not None:
            lines.append(f"pr_files: {session.get('pr_file_count')}")
        for path in (session.get("pr_files") or [])[:8]:
            lines.append(f"  {path}")
    lines.append(f"live: {'yes' if live else 'no'}")
    _append_well(lines, session, sid)
    if plan.strip():
        lines.extend(["", "Plan", "-----", plan.strip()])
    if events:
        lines.extend(["", "Recent", "------"])
        for ev in events[-12:]:
            kind = str(ev.get("kind") or "update").replace("_", " ")
            body = redact(str(ev.get("text") or ""))[:280]
            if body:
                lines.append(f"[{kind}] {body}")
    lines.extend(
        [
            "",
            "Continue this work. Do not assume the previous pager is still attached.",
        ]
    )
    if live:
        lines.append("Previous session is a live TUI. Do not inject into it.")
    inject_ok = (not live) and bool(acp_cwd)
    reason = None
    if live:
        reason = "live TUI pager; copy the pack, do not inject"
    elif not acp_cwd:
        reason = "no local clone for this project; copy the pack"
    return {
        "session_id": sid,
        "source": session.get("source") or "unknown",
        "title": redact(session.get("title") or sid),
        "cwd": cwd,
        "acp_cwd": acp_cwd,
        "branch": branch,
        "url": session.get("url"),
        "live": live,
        "inject_ok": inject_ok,
        "reason": reason,
        "text": "\n".join(lines),
    }


def main() -> int:
    if len(sys.argv) < 2:
        print("usage: handoff.py <session-id>", file=sys.stderr)
        return 2
    print(json.dumps(build_handoff(sys.argv[1]), indent=2, ensure_ascii=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src-tauri/src/handoff.rs
//! Cross-tool handoff packs. Never inject into a live TUI pager.

use crate::model::HandoffPack;
use crate::projects;
use crate::redact::{is_session_id, redact};
use crate::snapshot::{session_detail, tui_live_ids};
use crate::web::is_web_id;
use std::path::Path;
use std::process::Command;

pub fn build_handoff(id: &str) -> Result<HandoffPack, String> {
    if !(is_session_id(id) || is_web_id(id)) {
        return Err("invalid session id".into());
    }
    let detail = session_detail(id)?;
    let live = detail.session.live || tui_live_ids().contains(id);
    let cwd = detail.session.cwd.clone();
    let clone = projects::clone_path_for(&detail.session);
    let web = is_web_id(id);
    let acp_cwd = match &clone {
        Some(p) => p.clone(),
        None if web => String::new(),
        None => acp_cwd_for(&cwd),
    };
    let branch = git_head(if acp_cwd.is_empty() { &cwd } else { &acp_cwd });
    let mut lines = Vec::new();
    lines.push("ORBIT HANDOFF".into());
    lines.push(format!("source: {}", detail.session.source));
    lines.push(format!("id: {}", detail.session.id));
    lines.push(format!("title: {}", redact(&detail.session.title)));
    if let Some(pid) = &detail.session.project_id {
        lines.push(format!("project: {pid}"));
    }
    if !cwd.is_empty() {
        lines.push(format!("cwd: {cwd}"));
    }
    if !acp_cwd.is_empty() && acp_cwd != cwd {
        lines.push(format!("clone: {acp_cwd}"));
    }
    if let Some(r) = &detail.session.remote {
        lines.push(format!("remote: {r}"));
    }
    if let Some(b) = &branch {
        lines.push(format!("branch: {b}"));
    }
    if let Some(m) = &detail.session.model {
        lines.push(format!("model: {m}"));
    }
    if let Some(u) = &detail.session.url {
        lines.push(format!("url: {u}"));
    }
    if let Some(pr) = detail.session.pr_url.as_deref().filter(|u| !u.is_empty()) {
        let st = detail.session.pr_state.as_deref().unwrap_or("unknown");
        lines.push(format!("pr: {st} {pr}"));
        if let Some(n) = detail.session.pr_file_count {
            lines.push(format!("pr_files: {n}"));
        }
        for file in detail.session.pr_files.iter().take(8) {
            lines.push(format!("  {file}"));
        }
    }
    lines.push(format!("live: {}", if live { "yes" } else { "no" }));
    append_well_lines(&mut lines, &detail.session.id, detail.session.project_id.as_deref());
    if let Some(plan) = detail.plan_excerpt.as_deref().filter(|p| !p.trim().is_empty()) {
        lines.push(String::new());
        lines.push("Plan".into());
        lines.push("-----".into());
        lines.push(truncate(&redact(plan), 1200));
    }
    if !detail.events.is_empty() {
        lines.push(String::new());
        lines.push("Recent".into());
        lines.push("------".into());
        for ev in detail.events.iter().rev().take(12).collect::<Vec<_>>().into_iter().rev()
        {
            let kind = ev.kind.replace('_', " ");
            lines.push(format!("[{kind}] {}", truncate(&redact(&ev.text), 280)));
        }
    }
    lines.push(String::new());
    lines.push(
        "Continue this work. Do not assume the previous pager is still attached.".into(),
    );
    if live {
        lines.push("Previous session is a live TUI. Do not inject into it.".into());
    }
    let inject_ok = !live && !acp_cwd.is_empty();
    let reason = if live {
        Some("live TUI pager; copy the pack, do not inject".into())
    } else if acp_cwd.is_empty() {
        Some("no local clone for this project; copy the pack".into())
    } else {
        None
    };
    Ok(HandoffPack {
        session_id: detail.session.id,
        source: detail.session.source,
        title: redact(&detail.session.title),
        cwd,
        acp_cwd,
        branch,
        url: detail.session.url,
        live,
        inject_ok,
        reason,
        text: lines.join("\n"),
    })
}

fn append_well_lines(lines: &mut Vec<String>, sid: &str, project_id: Option<&str>) {
    let Some(pid) = project_id.filter(|p| projects::is_named_well(p)) else {
        return;
    };
    let snap = crate::snapshot::collect_snapshot();
    if let Some(desk) = snap.attention.iter().find(|a| {
        a.kind == "desk_claim" && a.title.to_ascii_lowercase().contains(&format!("claim {pid}"))
    }) {
        lines.push(format!("desk: {}", redact(&desk.title)));
    }
    let members: Vec<_> = snap
        .sessions
        .iter()
        .filter(|s| s.id != sid && s.project_id.as_deref() == Some(pid))
        .take(6)
        .collect();
    if members.is_empty() {
        return;
    }
    lines.push(String::new());
    lines.push("Well".into());
    lines.push("----".into());
    for session in members {
        lines.push(format!(
            "- {} ({})",
            redact(&session.title),
            projects::member_kind(session)
        ));
    }
}

fn acp_cwd_for(cwd: &str) -> String {
    if !cwd.is_empty() && Path::new(cwd).is_dir() {
        return cwd.to_string();
    }
    std::env::var("USERPROFILE")
        .or_else(|_| std::env::var("HOME"))
        .unwrap_or_default()
}

fn git_head(cwd: &str) -> Option<String> {
    if cwd.is_empty() || !Path::new(cwd).is_dir() {
        return None;
    }
    let branch = git_out(cwd, &["rev-parse", "--abbrev-ref", "HEAD"])?;
    let sha = git_out(cwd, &["rev-parse", "--short", "HEAD"])?;
    if branch.is_empty() {
        return None;
    }
    Some(format!("{branch} @ {sha}"))
}

fn git_out(cwd: &str, args: &[&str]) -> Option<String> {
    let out = Command::new("git").args(args).current_dir(cwd).output().ok()?;
    if !out.status.success() {
        return None;
    }
    let s = String::from_utf8_lossy(&out.stdout).trim().to_string();
    if s.is_empty() {
        None
    } else {
        Some(s)
    }
}

fn truncate(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        s.to_string()
    } else {
        let t: String = s.chars().take(max.saturating_sub(3)).collect();
        format!("{t}...")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::paths::grok_home;

    #[test]
    fn live_pack_refuses_inject() {
        let path = grok_home().join("active_sessions.json");
        let Ok(raw) = std::fs::read_to_string(&path) else {
            return;
        };
        let rows: Vec<serde_json::Value> = serde_json::from_str(&raw).unwrap_or_default();
        let Some(id) = rows.iter().find_map(|r| {
            let id = r.get("session_id").and_then(|v| v.as_str())?;
            let pid = r.get("pid").and_then(|v| v.as_u64())? as u32;
            if is_session_id(id) && crate::snapshot::tui_live_ids().contains(id) && pid > 0 {
                Some(id)
            } else {
                None
            }
        }) else {
            return;
        };
        let pack = build_handoff(id).expect("handoff");
        assert!(pack.text.contains("ORBIT HANDOFF"), "{}", pack.text);
        assert!(pack.text.contains("source:"));
        assert!(!pack.inject_ok, "{pack:?}");
        assert!(pack.live);
        assert!(!pack.text.contains("ghp_"));
    }

    #[test]
    fn package_redacts_tokens() {
        let dirty = redact("token ghp_exampleplaceholder000000 x");
        assert_eq!(dirty, "token [redacted] x");
    }
}
